### ISSUE TITLE
Snake-Case JSON Fields Ignored by UnmarshalKey()

### ISSUE THREAD

**Issue by coreywkruger** [spf13/viper#125]
I have encountered a situation where snake case fields returned from `UnmarshalKey()` are ignored while non snake case fields are returned normally. The snake case fields are returned as empty strings (the default) as if no value was found. NOTE: Json notation has been included on the struct as per usual. 

For an example of what I mean, execute the test program in this repo: https://github.com/coreywkruger/viper-test . Instructions are in README.md 

The json file that is being loaded is of the format:

``` json
{
  "data":{
      "normal": "Normal stuff",
      "snake_case": "Snake case stuff that is skipped :("
    }
}
```

The struct being populated is defined here:

``` go
type configJSON struct {
    Normal    string `json:"normal"`
    SnakeCase string `json:"snake_case"`
}
```

The json file is loaded like so:

``` go
// Unmarshalling from viper config:
config := viper.New()
config.SetConfigFile("./config.json")
err := config.ReadInConfig()
if err != nil {
    log.Fatal(err)
}
```

... then un-marshaled  and printed ...

``` go
c := configJSON{}
config.UnmarshalKey("data", &c)

// Print out the values of the fields
fmt.Printf("normal: `%s`\n", c.Normal)
fmt.Printf("snake_case: `%s`\n", c.SnakeCase) // <-- Notice how this is empty
```

The console output looks like:

```
normal: `Normal stuff`
snake_case: ``
```

`snake_case` is an empty string (but shouldn't be).

**Comment by fgrosse** [comment]
I encountered similar behavior when using `viper.Unmarshal`. The issue is that the value is first read via `viper.ReadInConfig` into the internal representation. The `Unmarshal` function then tries to apply this to a struct but the information that the original value was JSON encoded is no longer existent.

The [documentation of `Unmarshal`][1] kind of hints to the solution:
> // Unmarshals the config into a Struct. Make sure that the tags
> // on the fields of the structure are properly set.


.After some digging through the code I realized the correct default tag name is `mapstructure`
Assuming that `viper.UnmarshalKey` works like `viper.Unmarshal` you would need to define your struct like this in order to make it work you:

```golang
type configJSON struct {
    Normal    string `mapstructure:"normal"`
    SnakeCase string `mapstructure:"snake_case"`
}
```

Alternatively you could manually create the decoder and configure the tag yourself:

```
c := configJSON{}
config := DecoderConfig{	Result: &c, TagName: "json"}
decoder, err := NewDecoder(&config)
// handle err
err = decoder.Decode(viper.Get("data")
// handle err
```

I didnt have time to try this out so there might be typos there but the `mapstructure` approach worked fine for me.

I guess in the short term the configuration should be improved so the user knows the default value for those tags. In the long term it would be great if viper could remember the configuration format that was used in `viper.ReadInConfig` and configure the flags then automatically.

[1]: https://godoc.org/github.com/spf13/viper#Unmarshal

**Comment by trenton42** [comment]
I just ran into this, but mainly because I wasn't paying attention. @spf13 I think you can close this as it is [somewhat documented](https://github.com/spf13/viper#unmarshaling) in the readme (but perhaps you can call more attention to it there).

**Comment by oryband** [comment]
I just spent an hour struggling with this until I stumbled upon this issue.

1. Yes, this is documented somewhere deep in the long README. However, this is hard to find.
2. Please add this to the code doc of `Unmarshal()`, which will propagate into GoDoc as well. This would've saved me a lot of time.

In the future, I think mentioning stuff like this should be documented in the code, and only then in addition documented in README as well. A lot of people turn to the GoDoc first instead of the repo's README.

**Comment by oryband** [comment]
Also, this isn't really documented clearly in the README as well - it's just in the code example there. It's not explicitly explained anywhere.

**Comment by oryband** [comment]
And another thing: why do the struct tag key is `mapstructure`? This is a viper dependency but should be abstracted from the user. A better key name should be `viper`. This is also the convention with other libraries using struct tags such as https://github.com/asaskevich/govalidator

**Comment by trenton42** [comment]
@oryband I agree, and ran into the same issue as well. The wrapper around `mapstructure` is pretty thin, so it would probably take some work to abstract that out just to change the struct tags.

The documentation would be quick to change, though. Perhaps a PR?

**Comment by rskumar** [comment]
I am also getting same issue, is there any working hack or better solution?

**Comment by spf13** [comment]
Please send a PR. A doc update at the very least would help a lot of other
users.

**Comment by joe94** [comment]
Yeah, I wasted good 30 mins before I found this 😢

**Comment by Niondir** [comment]
Same here. Not very intuitive.

**Comment by andorus911** [comment]
I wasted an hour.
And another hour with confita because of the same problem. 😞
